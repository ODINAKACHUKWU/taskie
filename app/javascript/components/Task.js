import React, { useState, useEffect } from "react";
import { FaUserCircle } from "react-icons/fa";
import PropTypes from "prop-types";
import axios from "axios";
import { format } from "../helpers/functions";

function Task({ task }) {
  const [isChecked, setIsChecked] = useState(false);
  const [currentTask, setCurrentTask] = useState({});

  useEffect(() => {
    setCurrentTask(task);
  }, []);

  const handleChange = (e) => {
    setIsChecked(e.target.checked);
    axios
      .patch(`/api/v1/tasks/${task?.slug}`, { completed: e.target.checked })
      .then((res) => setCurrentTask(res.data))
      .catch((err) => console.log(err));
  };

  return (
    <div className="task-card">
      <div className="avatar-section">
        {task.avatar_url ? (
          <div className="avatar-container">
            <img src={task?.avatar_url} alt="Avatar" className="avatar" />
          </div>
        ) : (
          <FaUserCircle className="user-icon" />
        )}
      </div>
      <div className="task-details">
        <div className="task-title">{task?.title}</div>
        <div className="task-description">{task?.description}</div>
      </div>
      <div className="update-action">
        {currentTask.completed ? (
          <span className="time">{format(task?.updated_at)}</span>
        ) : (
          <input type="checkbox" checked={isChecked} onChange={handleChange} />
        )}
      </div>
    </div>
  );
}

Task.propTypes = {
  task: PropTypes.object.isRequired,
};

export default Task;
