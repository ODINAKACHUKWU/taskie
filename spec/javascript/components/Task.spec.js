import "jsdom-global/register";
import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Task from "../../../app/javascript/components/Task";

const task = {
  title: "Task 1",
  description: "This is task 1",
  avatar_url: "",
};

describe("rendering Task component", () => {
  it("renders Task component without crashing", () => {
    shallow(<Task task={task} />);
  });

  it("renders avatar section without crashing", () => {
    const wrapper = shallow(<Task task={task} />);
    const avatar = wrapper.find(".avatar-section");

    expect(avatar.exists()).toEqual(true);
    expect(avatar.type()).toEqual("div");
  });

  it("renders task details without crashing", () => {
    const wrapper = shallow(<Task task={task} />);
    const taskDetails = wrapper.find(".task-details");
    const taskTitle = wrapper.find(".task-title");
    const taskDescription = wrapper.find(".task-description");

    expect(taskDetails.exists()).toEqual(true);
    expect(taskDetails.type()).toEqual("div");
    expect(taskTitle.exists()).toEqual(true);
    expect(taskTitle.type()).toEqual("div");
    expect(taskDescription.exists()).toEqual(true);
    expect(taskDescription.type()).toEqual("div");
  });

  it("renders update action without crashing", () => {
    const wrapper = shallow(<Task task={task} />);
    const updateAction = wrapper.find(".update-action");

    expect(updateAction.exists()).toEqual(true);
    expect(updateAction.type()).toEqual("div");
  });
});

describe("passing props", () => {
  it("accepts task prop", () => {
    const wrapper = mount(<Task task={task} />);

    expect(wrapper.props().task).toEqual(task);
  });
});

describe("snapshot", () => {
  it("Task snapshot", () => {
    const tree = shallow(<Task task={task} />);

    expect(toJSON(tree)).toMatchSnapshot();
  });
});
