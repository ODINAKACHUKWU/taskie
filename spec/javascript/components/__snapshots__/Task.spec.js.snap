// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Task snapshot 1`] = `
<div
  className="task-card"
>
  <div
    className="avatar-section"
  >
    <FaUserCircle
      className="user-icon"
    />
  </div>
  <div
    className="task-details"
  >
    <div
      className="task-title"
    >
      Task 1
    </div>
    <div
      className="task-description"
    >
      This is task 1
    </div>
  </div>
  <div
    className="update-action"
  >
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
  </div>
</div>
`;
